crates/shared/src/code_simulation.rs

```diff
@@ -178,6 +178,39 @@ impl TryFrom<StateOverride> for StateObject {
     }
 }
 
+/// A code simulator that uses Web3 in the general case, but will create and
+/// save failed simulations on Tenderly to facilitate debugging.
+pub struct Web3ThenTenderly {
+    web3: Web3,
+    tenderly: TenderlyCodeSimulator,
+}
+
+impl Web3ThenTenderly {
+    pub fn new(web3: Web3, tenderly: TenderlyCodeSimulator) -> Self {
+        Self {
+            web3,
+            tenderly: tenderly.save(true, true),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl CodeSimulating for Web3ThenTenderly {
+    async fn simulate(
+        &self,
+        call: CallRequest,
+        overrides: StateOverrides,
+    ) -> Result<Vec<u8>, SimulationError> {
+        match self.web3.simulate(call.clone(), overrides.clone()).await {
+            Ok(value) => Ok(value),
+            Err(err) => {
+                tracing::warn!(?err, "web3 code simulation failed");
+                self.tenderly.simulate(call, overrides).await
```

Comment:
You could run the fallback tenderly simulation in a separate task to not delay the price estimation logic unnecessarily.

Reply:
The estimation wants the result though right? It can produce an error if this fails.

Reply:
Yes, but in the case where you use the `tenderly` simulator the `web3` simulator already produced an `Err`. Why not simply return that after spawning the `tenderly` simulation task? I wouldn't expect the `tenderly` simulation to work when the `web3` simulation already failed so we might as well return that result ASAP.